cri/v1alpha2/cri.go

```diff
@@ -248,65 +249,83 @@ func (c *CriManager) RunPodSandbox(ctx context.Context, r *runtime.RunPodSandbox
 		return nil, fmt.Errorf("failed to make sandbox pouch config for pod %q: %v", config.Metadata.Name, err)
 	}
 
-	sandboxName := makeSandboxName(config)
-
-	createResp, err := c.ContainerMgr.Create(ctx, sandboxName, createConfig)
+	// generate the id of sandbox meta and store it.
+	podSandboxID, err := c.generateSandboxID()
 	if err != nil {
-		return nil, fmt.Errorf("failed to create a sandbox for pod %q: %v", config.Metadata.Name, err)
+		return nil, err
 	}
-	id := createResp.ID
-
-	// once sandbox container created, we are obligated to store it.
 	sandboxMeta := &SandboxMeta{
-		ID:      id,
-		Config:  config,
-		Runtime: config.Annotations[anno.KubernetesRuntime],
+		ID: podSandboxID,
+	}
+	if err := c.SandboxStore.Put(sandboxMeta); err != nil {
+		return nil, err
 	}
 
+	// If running sandbox failed, clean up the container.
+	defer func() {
+		if !removeContainerErr {
+			// should not remove the sandbox container metadata from sandboxStore
+			// until it was removed by pouchd.
+			if err := c.SandboxStore.Remove(podSandboxID); err != nil {
+				logrus.Errorf("failed to remove the metadata of container %q from sandboxStore: %v", podSandboxID, err)
+			}
+		}
+	}()
+
 	if _, ok := config.Annotations[anno.LxcfsEnabled]; ok {
 		enableLxcfs, err := strconv.ParseBool(config.Annotations[anno.LxcfsEnabled])
 		if err != nil {
 			return nil, err
 		}
 		sandboxMeta.LxcfsEnabled = enableLxcfs
+
+		if err := c.SandboxStore.Put(sandboxMeta); err != nil {
+			return nil, err
+		}
+	}
+
+	sandboxName := makeSandboxName(config)
+
+	createResp, err := c.ContainerMgr.Create(ctx, sandboxName, createConfig)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create a sandbox for pod %q: %v", config.Metadata.Name, err)
 	}
+	sandboxContainerID := createResp.ID
 
+	// once sandbox container created, we are obligated to store it.
+	sandboxMeta.SandboxContainerID = sandboxContainerID
+	sandboxMeta.Config = config
+	sandboxMeta.Runtime = config.Annotations[anno.KubernetesRuntime]
```

Comment:
Just asking, Could the `Annotations` be nil?

Reply:
Yeah, and pouchd will check that.

Reply:
Does CRI check that?